File: backend/src/main/scala/services/battle/objects/apiTypes/state/BattleStateRootResponseApiTypes.scala
package services.battle.objects.apiTypes.state

import io.circe.{Encoder, Json}
import io.circe.syntax.*
import services.battle.objects.{BattleArtifactStatus, BattlePhase}
import services.battle.objects.core.BattleAggregateState

import BattleStateVectorResponse.given
import BattleStateProjectileResponse.given
import BattleStateProjectileTerminalResponse.given
import BattleStateSlowFieldResponse.given
import BattleStatePickupResponse.given
import BattleStateEventResponse.given
import BattleStatePlayerResponse.given

object BattleStateRootResponse {
  given Encoder[BattleAggregateState] =
    Encoder.instance(state =>
      Json
        .obj(
          "battleId" -> state.battleId.value.asJson,
          "roomId" -> state.roomId.value.asJson,
          "mapId" -> state.mapId.value.asJson,
          "phase" -> BattlePhase.wireValue(state.phase).asJson,
          "serverTime" -> state.serverTime.value.asJson,
          "startedAt" -> state.startedAt.value.asJson,
          "durationMs" -> state.durationMs.value.asJson,
          "elapsedMs" -> state.elapsedMs.value.asJson,
          "endsAt" -> state.endsAt.value.asJson,
          "worldSize" -> state.worldSize.asJson,
          "tick" -> state.tick.value.asJson,
          "resultReady" -> BattleArtifactStatus.isResultReady(state.artifactStatus).asJson,
          "replayReady" -> BattleArtifactStatus.isReplayReady(state.artifactStatus).asJson,
          "players" -> state.players.asJson,
          "projectiles" -> state.projectiles.asJson,
          "projectileTerminals" -> state.projectileTerminals.asJson,
          "slowFields" -> state.slowFields.asJson,
          "pickups" -> state.pickups.asJson,
          "events" -> state.events.asJson,
          "winnerPlayerId" -> state.winnerPlayerId.filter(_.value.trim.nonEmpty).map(_.value).asJson,
          "winnerHeroId" -> state.winnerHeroId.filter(_.value.trim.nonEmpty).map(_.value).asJson
        )
        .dropNullValues
    )
}

File: frontend/src/objects/battle/microservices/session/api/state/BattleStateRootResponseApiTypes.ts
import type { BattleExtractionResponseDto, BattleGasZoneResponseDto, BattleLootCacheResponseDto } from "../../../extraction/api/state/BattleExtractionStateResponseApiTypes";
import type { BattlePhase } from "../../../../objects/core/BattleCoreScalars";
import type { BattleStateEventResponseDto, BattleStatePickupResponseDto, BattleStateSlowFieldResponseDto } from "./BattleStateEntityResponseApiTypes";
import type { BattleStatePlayerResponseDto } from "./BattleStatePlayerResponseApiTypes";
import type { BattleStateProjectileResponseDto, BattleStateProjectileTerminalResponseDto } from "./BattleStateProjectileResponseApiTypes";
import type { BattleApiVectorDto } from "./BattleStateSharedResponseApiTypes";

export type AuthoritativeBattlePhaseDto = BattlePhase;

export interface BattleStateResponseDto {
  battleId: string;
  roomId: string;
  mapId: string;
  phase: AuthoritativeBattlePhaseDto;
  serverTime: number;
  startedAt: number;
  durationMs: number;
  elapsedMs: number;
  endsAt: number;
  worldSize: BattleApiVectorDto;
  tick: number;
  resultReady: boolean;
  replayReady: boolean;
  players: BattleStatePlayerResponseDto[];
  projectiles: BattleStateProjectileResponseDto[];
  projectileTerminals: BattleStateProjectileTerminalResponseDto[];
  slowFields: BattleStateSlowFieldResponseDto[];
  pickups: BattleStatePickupResponseDto[];
  gasZone?: BattleGasZoneResponseDto | null;
  extraction?: BattleExtractionResponseDto | null;
  lootCaches?: BattleLootCacheResponseDto[];
  events: BattleStateEventResponseDto[];
  winnerPlayerId?: string;
  winnerHeroId?: string;
}

